dremio_toolkit/env_diff.py

```diff
@@ -0,0 +1,269 @@
+#########################################################################
+# Copyright (C) 2023 UCE Systems Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Contact dremio@ucesys.com
+#########################################################################
+
+from dremio_toolkit.logger import Logger
+from dremio_toolkit.env_definition import EnvDefinition
+import os
+import json
+
+
+class DiffType:
+    MATCH = 0
+    DIFF_UID = 1
+    DIFF_ATTRIBUTE = 2
+    MISSING_ATTRIBUTE = 3
+    DIFF_PERMISSIONS = 4
+    DIFF_OWNER = 5
+
+    @classmethod
+    def all(cls) -> list:
+        return [cls.MATCH, cls.DIFF_UID, cls.DIFF_ATTRIBUTE]
+
+
+class EnvDiff:
+    diff_containers = []
+    diff_sources = []
+    diff_spaces = []
+    diff_folders = []
+    diff_vds = []
+    diff_reflections = []
+    diff_rules = []
+    diff_queues = []
+    diff_tags = []
+    diff_wikis = []
+
+    _base_def: EnvDefinition
+    _comp_def: EnvDefinition
+
+    def __init__(self, logger: Logger):
+        self._logger = logger
+
+    def diff_snapshot(self, base_env_def: EnvDefinition, comp_env_def: EnvDefinition) -> None:
+        self._base_def = base_env_def
+        self._comp_def = comp_env_def
+        self._logger.new_process_status(100,'Comparing snapshots. ')
+        self._diff_containers()
+        self._logger.print_process_status(complete=5)
+        self._diff_sources()
+        self._logger.print_process_status(complete=10)
+        self._diff_spaces()
+        self._logger.print_process_status(complete=15)
+        self._diff_folders()
+        self._logger.print_process_status(complete=30)
+        self._diff_vds()
+        self._logger.print_process_status(complete=80)
+        self._diff_reflections()
+        self._logger.print_process_status(complete=85)
+        self._diff_rules()
+        self._diff_queues()
+        self._logger.print_process_status(complete=90)
+        self._diff_tags()
+        self._logger.print_process_status(complete=95)
+        self._diff_wikis()
+        self._logger.print_process_status(complete=100)
+
+    def write_diff_report(self, filename: str) -> None:
+        if os.path.isfile(filename):
+            os.remove(filename)
+        diff_json = {
+            "diff": [
+                {"base": self._base_def.endpoint},
```

Comment:
base & comp are not very intuitive, how about `base_endpoint` and `comp_endpoint`?

Reply:
Applied. Thank you.